xline/src/storage/db.rs

```diff
@@ -1,141 +1,84 @@
-use parking_lot::{RwLock, RwLockUpgradableReadGuard};
-use prost::Message;
+use std::sync::Arc;
 
-use super::{storage_api::StorageApi, ExecuteError, Revision};
-use crate::rpc::{KeyValue, PbLease};
+use engine::{engine_api::StorageEngine, Delete, Put, WriteOperation};
+
+use super::{storage_api::StorageApi, ExecuteError};
 
 /// Database to store revision to kv mapping
-#[derive(Debug)]
-pub(crate) struct DB<S>
-where
-    S: StorageApi,
-{
+#[derive(Debug, Clone)]
+pub struct DB<S: StorageEngine> {
     /// internal storage of `DB`
-    storage: RwLock<S>,
+    engine: Arc<S>,
 }
 
 impl<S> DB<S>
 where
-    S: StorageApi,
+    S: StorageEngine,
 {
     /// New `DB`
-    pub(crate) fn new(storage: S) -> Self {
+    #[allow(dead_code)]
+    #[inline]
+    #[must_use]
+    pub fn new(engine: S) -> Self {
         Self {
-            storage: RwLock::new(storage),
+            engine: Arc::new(engine),
         }
     }
+}
+
+impl<S> StorageApi for DB<S>
+where
+    S: StorageEngine,
+{
+    fn create_table(&self, table: &str) -> Result<(), ExecuteError> {
+        self.engine.create_table(table).map_err(|e| {
+            ExecuteError::DbError(format!("Failed to create a new table {table}, error: {e}"))
+        })?;
+        Ok(())
+    }
 
-    /// Insert a `KeyValue`
-    pub(crate) fn insert(&self, revision: Revision, kv: &KeyValue) -> Result<(), ExecuteError> {
-        let mut storage = self.storage.write();
-        let key = revision.encode_to_vec();
-        let value = kv.encode_to_vec();
-        let _prev_val = storage.insert(key, value).map_err(|e| {
+    fn insert<K, V>(&self, table: &str, key: K, value: V) -> Result<(), ExecuteError>
+    where
+        K: AsRef<[u8]> + std::fmt::Debug,
```

Comment:
I this K and V should be `Into<Vec<u8>>` at here, otherwise there will be unnecessary heap allocation

Reply:
Fixed